cmd/pebble/cmd_ls.go

```diff
@@ -0,0 +1,106 @@
+// Copyright (c) 2022 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package main
+
+import (
+	"errors"
+	"fmt"
+	pathpkg "path"
+	"strings"
+
+	"github.com/jessevdk/go-flags"
+
+	"github.com/canonical/pebble/client"
+)
+
+// ErrPatternOutsideFileName is returned if a pattern is found anywhere but the file name of a path
+var ErrPatternOutsideFileName = errors.New("patterns can only be applied to file names, not directories")
+
+type cmdLs struct {
+	clientMixin
+	timeMixin
+
+	Directory  bool `short:"d" long:"directory"`
+	LongFormat bool `short:"l"`
+
+	Positional struct {
+		Path string `positional-arg-name:"<path>"`
+	} `positional-args:"yes" required:"yes"`
+}
+
+var lsDescs = map[string]string{
+	"directory": `Display information about the file system entry, instead of listing directory contents.`,
+	"l":         `Display file system entries in a list format.`,
+}
+
+var shortLsHelp = "List path contents"
+var longLsHelp = `
+The ls command takes a path (a file, directory or glob) and obtains its
+contents.
+`
+
+func (cmd *cmdLs) Execute(args []string) error {
+	if len(args) > 0 {
+		return ErrExtraArgs
+	}
+
+	path, pattern, err := parseGlob(cmd.Positional.Path)
+	if err != nil {
+		return err
+	}
+
+	files, err := cmd.client.ListFiles(&client.ListFilesOptions{
+		Path:    path,
+		Pattern: pattern,
+		Itself:  cmd.Directory,
+	})
+	if err != nil {
+		return err
+	}
+
+	w := tabWriter()
+	defer w.Flush()
+	for _, fi := range files {
+		if cmd.LongFormat {
+			fmt.Fprintf(w, "%s\t%s\t%s\t%s\t%s\n", fi.Mode().String(), fi.User(), fi.Group(), cmd.fmtTime(fi.ModTime()), fi.Name())
```

Comment:
I think we should add the size before the date column for type=file (probably make it "-" for other types).